onmt/utils/Cuda.lua

```diff
@@ -52,18 +52,23 @@ end
   When using CPU only, converts to float instead of the default double.
 ]]
 function Cuda.convert(obj)
-  if torch.typename(obj) then
+  local objtype = torch.typename(obj)
+  if objtype then
     if Cuda.activated and obj.cuda ~= nil then
-      return obj:cuda()
+      if objtype:find('torch%..*LongTensor') then
+        return obj:cudaLong()
+      else
+        return obj:cuda()
+      end
     elseif not Cuda.activated and obj.float ~= nil then
       -- Defaults to float instead of double.
       return obj:float()
     end
-  end
-
-  if torch.typename(obj) or type(obj) == 'table' then
-    for k, v in pairs(obj) do
-      obj[k] = Cuda.convert(v)
+  else
```

Comment:
I have already make the changes you asked to, but I still think this part of code could be merged to the first part, at least the repeated call to typename is unnecessary.